### PR TITLE
feat(forge/test): compile in sparse mode in forge test

### PR DESCRIPTION
**Pull request by mattsse**
## Motivation
sparse mode was recently introduced in ethers-rs https://github.com/gakonst/ethers-rs/pull/1039

with `Project::compile_sparse(filter)` only the files matching the specified filter are the primary target of the Solc call. Files, those that do not match the filter, but still need to be compiled (imports etc...), will have their compiler output pruned.

In other words, we explicitly tell solc to include artifact data (bytecode, abi, etc.) only for the files that the filter applies to, the output for all other files is explicitly set to empty.

## Solution

The used filter for forge test in sparse mode is
* --match-path <> if the argument was provided
* file ends with `.t.sol` otherwise

### CODE REVIEW

File: cli/src/compile.rs [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+    pub fn compile(self, project: &Project) -> eyre::Result<ProjectCompileOutput> {
+        self.compile_with(project, |prj| Ok(prj.compile()?))
+    }
+
+    /// Compiles the project with [`Project::compile_parse()`] and the given filter.
+    ///
+    /// This will emit artifacts only for files that match the given filter.
+    /// Files that do _not_ match the filter are given a pruned output selection and do not generate
+    /// artifacts.
+    pub fn compile_parse<F: FileFilter + 'static>(
```
Comment (onbjerg) [inline cli/src/compile.rs:46]: `compile_sparse`?

Comment (mattsse) [reply cli/src/compile.rs]: yep!

**Comment by gakonst** [comment]
I'm interested in benchmarking this a bit before merging, but really excited about this change

**Comment by gakonst** [comment]
Hmm I tried running Solmate tests with it but didn't seem to see any difference in time spent during the compilation test. I did "forge test -m TestWETH

**Comment by mattsse** [comment]
I added tracing to track time spent compiling the project.

on clean solmate with sparse it takes around ~11.5-12.5 secs on my machine, but can vary a bit

I've instrumented master with the same traces and `forge test` takes a bit longer on average, but also fluctuates.

Testing a single file with `--mp` is now significantly faster because we're only compiling this file

sparse:

```console
forge test --mp /Users/Matthias/git/rust/foundry-integration-tests/testdata/solmate/src/test/WETH.t.sol
Mar 21 15:00:12.491 TRACE forge_compile: start compiling project
[⠢] Compiling...
[⠘] Compiling 8 files with 0.8.10
Mar 21 15:00:13.317 TRACE forge_compile: finished compiling after 825.26039ms
```

master:

```console
forge test --mp /Users/Matthias/git/rust/foundry-integration-tests/testdata/solmate/src/test/WETH.t.sol
Mar 21 15:00:36.866 TRACE forge_compile: start compiling project
[⠆] Compiling...
[⠊] Compiling 53 files with 0.8.10
Mar 21 15:00:50.336 TRACE forge_compile: finished compiling after 13.47044555s
```

**Review by gakonst — CHANGES_REQUESTED** [review]
Supportive of merging, we should also do a parsing step with solang so that we can filter when -m is used and not just with --mp

**Pull request merged**